Fix Form publishing

dave-mills opened the pull request:
fixes #298. 

This PR has 2 main updates:


### Main fix

1. Adjusts the "Publish form" function on the XlsformsTableView so that it runs the correct publishing function. Previously, it re-ran the deployDraft() method, but did not then proceed to the publishing. 

This publishes the form, regardless of the status. This _should_ be fine, as there should always be a draft on ODK Central (every time the form is published, a new draft is created immediately afterwards). 


### Dev helper function

2. Updates the `php artisan app:reset-real` and `app:reset-mini` commands, so that as well as pulling any updates to the xlsform_versions table present on ODK Central, it also updates the xlsforms table, specifically adding the entries for :
- enketo_draft_id
- enketo_id
- is_active
- odk_draft_updated_at
- odk_published_at

Based on these properties, the forms now display the correct status in the PilotIndex and SetUpSurvey pages immediately after doing the database reset. 

### Notes for testing:

@dan-tang-ssd: 

Using the 'real' forms, On the ODK Central server I've left it with the following setup:

| Team | Form | Status | 
| -- | -- | -- |
| P1 Test Team | HOLPA Household Survey | Live | 
| P1 Test Team | HOLPA Fieldwork Survey | not yet published (will show as "INACTIVE" on the platform) | 
| P2 Test Team | HOLPA Household Survey | Live | 
| P2 Test Team | HOLPA Fieldwork Survey |  Live |
| Non Program Test Team | HOLPA Household Survey | not yet published |
| Non Program Test Team | HOLPA Fieldwork Survey | not yet published | 

After running `php artisan app:reset-real`, this should be reflected in the platform / database. 


For reference:

- P1 Test Team on ODK Central: https://odk-test.stats4sdtest.online/projects/2728
- P2 Test Team: https://odk-test.stats4sdtest.online/projects/2729
- Non Program Test Team: https://odk-test.stats4sdtest.online/projects/2730

---

To test the publishing, I suggest: 

- confirm that the app:reset-real works to give you the state above;
- publish one of the not-yet-published forms, and confirm it's published and available on ODK Central
- make a change to the survey (not all the changes are currently correctly marking the form as "draft_needs_update", but updating the choice lists does work. Try adding a new language entry in the **Survey Dashboard > 
Localisation: Place-based adaptations > Contextualise choice lists page.** 
- Check that change is pushed to ODK Central in the latest draft format. (The language choices are the 1st question in the household form) - it's quickest to test with Enketo in the browser.
- Publish the form through the platform.
- Confirm that the now published form includes the new language choice list entry.

dave-mills commented:
Thanks Dan! Looks mostly good and as expected. One comment on your step-by-step test: 

> yes, the command give me the state as above at the first time. But after publishing an INACTIVE form. This published form keeps "published" even I run the command again.

Good - that's expected. Resetting the local database doesn't impact the ODK Central server, so the published form will remain published. Good that the reset:real scripts now pull that published status and show it on the local platform correctly. 


### Memory Allocation Error

Ooh, interesting. This makes sense - dealing with the large HOLPA xlsforms is probably one of the bigger tasks we do in terms of putting raw data into memory for processing. For a local fix, it looks like you can adjust the `sort_buffer_size` variable in your local MySQL config: https://philipmcclarence.com/solving-mysql-error-1038-sqlstate-hy001-er_out_of_sortmemory-increasing-server-sort-buffer-size/

Longer term, we should consider:
- stress testing this part on the server
- potential optimisations to reduce memory load when running back-to-back jobs
- testing the fallbacks - e.g. what happens when a job fails like this in production? Is our job queue settings sufficient to ensure the jobs retry after an appropriate time, or send the user a notification if it fails permanently?

dan-tang-ssd commented:
I run command "php artisan app:reset-real" to refresh my local database.

P1 Test Team. 
Languages: English, French, Spanish in application
Same languages in household form opened in Enketo

---

Screen shots:

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/16a299ac-3b04-487c-b54e-a612c48d38ef" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/e09cc21a-6162-4119-992c-8c75af54d3c5" />

dave-mills commented:
I've pushed an update to the form publishing workflow: 


1. Now, the `Xlsform` class's `publishForm()` method returns a PendingDispatch for a job, instead of doing the publishing directly. This is because we need to check whether there is _also_ a draft update required (i.e. `draft_needs_update` is true). If there is, we return the Draft job as a queued item, and chain on the new `PublishXlsformOnOdkCentral` job class to run after the draft is deployed. 

2. Because the form publishing is now a server-side queued process, we can't just immediately return the success or fail response. Instead, the immediate notification is now a "publishing has started", and I've added listeneers to the XlsformTableView Livewire component to listen for the appropriate events that trigger when the forms are published.

3. I've updated the flow so that _after_ publishing, a new draft is deployed. However, we no longer trigger a recreation of the Xlsform - because we know no changes have been made. So we can skip the $xlsform->generateXlsfile() and save a lot of time.

I also tweaked the UI slightly to show a notice on the 'status' table column when a form update is available to publish.


I'd suggest reviewing this setup both where:

- `draft_needs_update` *and* `live_needs_update* is true.
- `draft_needs_update` is false, but `live_needs_update` is true.

dave-mills commented:
Reference for the UI update: 

<img width="1652" height="1357" alt="Sceenshot 2025-12-16 at 16 26 30" src="https://github.com/user-attachments/assets/26deab33-1f7d-41f6-84dd-68b42c1003dd" />

dan-tang-ssd commented:
I added the 5th option for languages.

In "initial pilot" page, household form QR code not showed. It keeps loading.

In main pilot page, I click "PUBLISH CHANGES" button for Household form.

In Enketo, all five options showed.

---

Screen shots:

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/b404f837-b751-488d-a233-640081915106" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/6f753628-7b80-4ccc-b3cd-7da3958a97f4" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/63a320c2-c5f8-4938-b578-a853c6dfe664" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/40af863e-27c1-42a0-b359-2ede0667bb7e" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/cd173550-c16f-438a-9931-f426dbe5958f" />

dave-mills commented:
> Then I go to "initial pilot" page.
> The QR code for household form keeps loading... Is it normal...?

It will continue loading until the page hears the `XlsformDraftWasDeployed` event from the server; when it refreshes the livewire $xlsform variable. If it continues loading after that event is dispatched, it might be that you don't have Reverb running or something similar.  A page refresh when `draft_needs_update===0` should do it.

dan-tang-ssd commented:
I added the 7th option.

draft_needs_update and live_needs_update are both 1.

I do not visit "initial pilot" page.

Then I go to main pilot page to publish household form.

In Enketo, household form shows 6 options.

---

Screen shots:

Executed job for the 7th option
<img width="881" height="481" alt="image" src="https://github.com/user-attachments/assets/04533419-87c9-4400-abbe-1d84d572e997" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/1bb0ccd6-0694-4ff5-a48e-38905f1a2847" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/b47eae21-52a8-4eb8-99f6-1b4abe88ea75" />

dave-mills commented:
@dan-tang-ssd - my question remains - you should be seeing the new publish xlsform job being run; the fact that you are not suggests that you do not have my latest update.

dan-tang-ssd commented:
> I don't see the publish job. Are you sure you have the latest update that includes the new `PublishXlsformOnOdkCentral` job class?

I did a search for "PublishXlsformOnOdkCentral" in VS Code, I cannot find this class.

I run command "git pull" to pull latest code from "fix-xlsform-publishing" branch.
Do I need to run command "git submodule update" to get latest code from submdoules?

dan-tang-ssd commented:
I run command "git submodule update" to get latest code from submodule.
Now I can find "PublishXlsformOnOdkCentral" in VS code.

I added the 8th option.
draft_needs_update and live_needs_update are both 1.
Then I go to main pilot page to publish the household form.

In Enketo, household form shows eight options.

---

Screen shots:

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/8805a817-c777-4f1e-82c0-e1cbd10d0713" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/b58fcd92-1e39-4ad5-bbfe-406b36954ced" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/86732b68-dad3-4907-98ae-2fa194f1cd10" />

dan-tang-ssd commented:
> > draft_needs_update is false, but live_needs_update is true.
> 
> Question: Is there anything I can update but keep draft_needs_update as false? So that I can verify the household form has been published after publishing from application?

I reset my local database to test this case.

Only 3 options for languages.

draft_needs_update is false, but live_needs_update is true.

Checked household form in Enketo, it has 8 options for languages now.

I go to main pilot page to publish household form.

Error occurred when running task DeployDraftXlsformToOdkCentral
2025-12-17 08:52:21 Stats4sd\FilamentOdkLink\Jobs\XlsformDeployment\DeployDraftXlsformToOdkCentral  247.16ms FAIL

---

Screen shots:

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/ceb258c1-4936-4eb5-9493-f3d853365395" />

<img width="1006" height="274" alt="image" src="https://github.com/user-attachments/assets/489cd295-04f6-483e-9edc-1cebc183590f" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/3d05166e-4642-453d-adaa-dac80c33d901" />

<img width="1920" height="982" alt="image" src="https://github.com/user-attachments/assets/b84c4e53-f777-49b6-96cd-6477ad835ac7" />

<img width="1010" height="244" alt="image" src="https://github.com/user-attachments/assets/851d825f-b5c1-4004-9e76-6f3b2bf2f2d9" />

---

Error details in Ray

<img width="1213" height="1011" alt="image" src="https://github.com/user-attachments/assets/432f65ae-236a-40ee-a181-bd7cf0ccc431" />

<img width="1213" height="1011" alt="image" src="https://github.com/user-attachments/assets/b3a8c1ae-db59-49d7-bcb1-0fdb161ed14f" />

<img width="1213" height="1011" alt="image" src="https://github.com/user-attachments/assets/443dd4a0-4f87-4b18-b1e6-a05670ee1433" />

<img width="1213" height="1011" alt="image" src="https://github.com/user-attachments/assets/31c9a7da-96bd-4f34-8d4b-0b0c5af1f555" />

dan-tang-ssd commented:
I added the 4th option for languages.

live_need_updates is false now. Should it be true instead?

I tried to update it manually in backend database, the "PUBLISH CHANGES" button in main pilot page will be showed.

---

Screen shots:

<img width="746" height="254" alt="image" src="https://github.com/user-attachments/assets/22a69ff9-39ce-4929-bf12-ed272a4233ff" />

dan-tang-ssd commented:
Hi @dave-mills 

I think the testing result is good.
Now the household form can be published to ODK central after adding a new option for languages.

One remaining item is to consider to set live_needs_update to false initially.

After sorting out this I think it is good to merge this PR.

dave-mills commented:
Agreed - lets set live-needs-update to false initially in the seeder files. That keeps the system more in sync with the ODK Central setup 👍. Just done that - will now merge this back in :)
